feat(terminal): close shell terminals automatically

gpanders opened the pull request:
When starting a terminal buffer with no arguments (i.e. `:terminal`), automatically close the buffer if the shell exits without an error.

Closes #8975.

gpanders committed:
feat(terminal): close shell terminals automatically

When starting a terminal buffer that uses the default value of &shell
(e.g. when using `:terminal` or `termopen([&shell])`), automatically
close the buffer if the shell exits without an error.

gpanders commented:
I'm going to see if I can also find a way to close https://github.com/neovim/neovim/issues/5176 as well.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9504,6 +9504,13 @@ static void ex_terminal(exarg_T *eap)
+    do_cmdline_cmd("autocmd TermClose <buffer> "
+                   " if !v:event.status | "
+                   "   exec 'bdelete! ' .. expand('<abuf>') | "
+                   " endif");
```
Probably should be global rather than `<buffer>` local. And we can declare it in a `nvim_` augroup so that users can opt-out.

gpanders replied:
I don’t think it should be global as then it would apply to all `termopen` calls rather than just `:terminal`. The former provides finer grained control so we should make fewer assumptions and allow the user to do whatever they want, while `:terminal` is a convenience function that should have more convenient defaults.

My 2¢.

justinmk replied:
Hmm. Thinking about how a user could opt-out of this in its current form: `autocmd TermClose` is done after the `do_cmdline_cmd(ex_cmd);` above. So this is setup after TermOpen... how would users opt-out?

> I don’t think it should be global as then it would apply to all `termopen` calls

Makes sense, but OTOH consistent behavior is easier to reason about, and to document. Clear documentation (one-liner that can be pasted), somewhat addresses the termopen() case, and avoids contortions for the :terminal case.

A global autocmd or augroup solves the user config case, but doesn't solve the termopen() case for scripts:  can't selectively opt-out, except by `:noautocmd call termopen(...)` ... maybe that's reasonable...

gpanders replied:
> A global autocmd or augroup solves the user config case, but doesn't solve the termopen() case for scripts: can't selectively opt-out, except by `:noautocmd call termopen(...)` ... maybe that's reasonable...

I think that's reasonable, especially if it's documented.

If we do make it a global autocommand then we may want to resolve https://github.com/neovim/neovim/pull/15401 first so that we can use whatever default autocommand framework we decide on.

gpanders replied:
I've been thinking about this more and two things occurred to me:

1) Making it a global autocommand means we can't use the heuristic of whether or not an argument was provided to `:terminal` to determine if we should automatically close the buffer or not. Right now, we create this autocommand if there is no argument and we are therefore using a default shell. But if, for instance, a user runs `:terminal ls` then we *do not* want to automatically close the buffer when the job completes. If we use a global autocommand we will need another flag or something to indicate which state we are in.

2) I wonder if it's not actually more common that users may want to opt-out of the auto-close behavior on a per-job/per-terminal instance, rather than opting out globally. This also provides plugins finer grained control to opt out of this behavior, which wouldn't be possible (or at least, more complex) with a global autocommand.

For these two reasons I think we ought to stick with a buffer-local autocommand. What do you think?

justinmk replied:
> 1\. ... But if, for instance, a user runs `:terminal ls` then we _do not_ want to automatically close the buffer when the job completes. If we use a global autocommand we will need another flag or something to indicate which state we are in.

yeah, that's where I was going with [this comment](https://github.com/neovim/neovim/pull/15440#discussion_r693233441):

> In a global TermOpen autocmd, that could possibly be derived from the `term://` buffer name, or maybe we should expose the args in `v:event`. Or `nvim_get_chan_info(&channel)`

re (2):

> This is much more straightforward to do with a buffer-local autocommand. This would also provide plugins finer-grained control to opt out of this behavior,

Should work, since this default autocmd will always be defined before any user/plugin autocmds. Just needs a unique augroup, something like `nvim_terminal_close`

gpanders replied:
> Should work, since this default autocmd will always be defined before any user/plugin autocmds. Just needs a unique augroup, something like `nvim_terminal_close`.

The plugin would need to disable the global autocommand in that case though, which the user may not want. The only way around this would be to add some kind of buffer-local variable that acts as a flag (e.g. `let b:terminal_autoclose = v:false`) that the plugin could set. But if we're doing that, why not just use a buffer-local autocommand?

I guess I disagree that a global autocommand is less messy, it paints with too broad of a brush so we end up needing to introduce further measures to account for edge cases and exceptions.

justinmk replied:
>> Should work, since this default autocmd will always be defined before any user/plugin autocmds. Just needs a unique augroup, something like `nvim_terminal_close`.

^ this was regarding buffer-local autocommand

gpanders replied:
Ah cool, sorry for the misunderstanding 😅

gpanders commented:
> I'm going to see if I can also find a way to close #5176 as well.

https://github.com/neovim/neovim/pull/15442 takes care of #5176, so if we can find a way to make the `TermClose` autocommand somehow enter the `terminal_enter` code path we can kill two birds with one stone.

justinmk commented:
Different idea: what if we only set this on `TermEnter ` (terminal buffer version of "InsertEnter")? I'm trying this in my config:

```
autocmd TermOpen *
  \ autocmd TermEnter <buffer=abuf> ++once
  \ autocmd TermClose <buffer=abuf> if !v:event.status | exe 'bdelete '..expand('<abuf>') | endif
```

gpanders commented:
> Different idea: what if we only set this on `TermEnter ` (terminal buffer version of "InsertEnter")? I'm trying this in my config:
> 
> ```
> autocmd TermOpen *
>   \ autocmd TermEnter <buffer=abuf> ++once
>   \ autocmd TermClose <buffer=abuf> if !v:event.status | exe 'bdelete '..expand('<abuf>') | endif
> ```
> 
> So far it works as desired with these cases:
> 
>     * `:te top` (interactive)
> 
>     * `:te ls` (non-interactive)
> 
>     * `:te` (default shell)

This won't work if the user uses `autocmd TermOpen * startinsert` (which the docs mention in `:h :terminal`). I'd wager this is not that uncommon.

justinmk commented:
> This won't work if the user uses `autocmd TermOpen * startinsert` (which the docs mention in `:h :terminal`). I'd wager this is not that uncommon.

good catch. At this point, it seems like this is a requirement:

> .... expose the `:terminal` args in `v:event`. Or `nvim_get_chan_info(&channel)`

gpanders commented:
> .... expose the :terminal args in v:event. Or nvim_get_chan_info(&channel)

Fortunately, `f_termopen()` already parses out the command argument; we just need to store that somewhere.

I'm thinking of adding a new `const char *cmd` field to `struct Terminal`. Do you have a better suggestion?

I've already looked at just using `proc->argv[0]`, but this memory is freed when the process exits, which can happen even if the terminal buffer is still alive (that wouldn't really be that helpful anyway, because in the event someone uses e.g. `:terminal ls`, `proc->argv[0]` would still just be the shell).

justinmk commented:
> I'm thinking of adding a new `const char *cmd` field to `struct Terminal`. Do you have a better suggestion?

`nvim_get_chan_info()` could include an `argv` field. PR: https://github.com/neovim/neovim/pull/15537 . Only works for running jobs currently, because...

> I've already looked at just using `proc->argv[0]`, but this memory is freed when the process exits, which can happen even if the terminal buffer is still alive

Yeah good point. I wonder if we can just link job lifetime to the channel lifetime. Is there a reason we need to free jobs before channels? Probably not...

> **EDIT:** Wait wait, if we're not doing the global autocommand, isn't it just sufficient to check `*eap->arg == NUL`?

Ideally we avoid using internal magic, so that these decisions can be made by scripts.

gpanders commented:
With #15537 the check simply becomes

```vim
if nvim_get_chan_info(&channel).argv ==# [&shell]
```

gpanders committed:
test: update tests for auto-closing :term buffers

The terminal buffer closes automatically when using `:terminal` and the
command exits without an error. This messes up some tests that expect
the terminal buffer to still be open. We can force the buffer not to
close by passing an argument to `:terminal`. This can be anything, since
the shell-test stub simply prints whatever argument it's given.

gpanders committed:
fix: handle argv not present in nvim_get_chan_info

gpanders commented:
Lint failure is on `screen.c` (why?), otherwise everything looks good!

clason commented:
> Lint failure is on `screen.c` (why?), otherwise everything looks good.

Because dundargoc taught uncrustify new tricks, tested them on `screen.c`, and nobody told the linter about it yet ;)

TL;DR: unrelated

zeertzjq commented:
Note: Because of #16257, this solution leaks memory with every terminal closed with `:bdelete` on `TermClose`.

gpanders committed:
Use Lua autocommand and make TermClose autocommand global

justinmk reviewed (approved):
Very happy to see this! We can continue to improve little details, but this looks like a monotonic improvement meanwhile.

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1107,13 +1107,26 @@ end
+        local info = vim.api.nvim_get_chan_info(vim.bo[args.buf].channel)
+        local argv = info.argv or {}
+        if #argv == 1 and argv[1] == vim.o.shell then
+          vim.cmd({ cmd = 'bdelete', args = { args.buf }, bang = true })
```
This will close the current split window (if any). That's probably unwanted in many cases, so we need either of:

- open `:terminal` in split by default
- add a `:bdelete` variant that doesn't close the current split (or hack it here using `b#`) https://github.com/neovim/neovim/issues/2434

gpanders replied:
I made https://github.com/neovim/neovim/pull/15442 a while ago for the first option. Even though it would probably break a lot of things, I think I’d still prefer the split-by-default behavior 😬

justinmk replied:
Do you mean https://github.com/neovim/neovim/pull/15427 ? Unless it breaks fugitive I think we should try it.

gpanders replied:
Oops, yes, thanks. Updated my original comment.
